src/pushprocessor.ts

```diff
@@ -112,6 +151,60 @@ const DEFAULT_UNDERRIDE_RULES: IPushRule[] = [
     },
 ];
 
+const EXPECTED_DEFAULT_UNDERRIDE_RULE_IDS = [
+    RuleId.IncomingCall,
+    RuleId.EncryptedDM,
+    RuleId.DM,
+    RuleId.Message,
+    RuleId.EncryptedMessage,
+];
+
+function mergeRulesWithDefaults(
+    incomingRules: IPushRule[],
+    defaultRules: IPushRule[],
+    defaultRuleIds: string[],
+): IPushRule[] {
+    let firstCustomRuleIndex = incomingRules.findIndex((r) => !r.default);
+    if (firstCustomRuleIndex < 0) firstCustomRuleIndex = incomingRules.length;
+
+    let nextExpectedRuleIdIndex = 0;
+    const newRules: IPushRule[] = [];
+    for (const rule of incomingRules.slice(0, firstCustomRuleIndex)) {
+        const ruleIndex = defaultRuleIds.indexOf(rule.rule_id);
+        if (ruleIndex === -1) {
+            // an unrecognised rule; copy it over
+            newRules.push(rule);
+            continue;
+        }
+        while (ruleIndex > nextExpectedRuleIdIndex) {
+            // insert new rules
+            const defaultRuleId = defaultRuleIds[nextExpectedRuleIdIndex];
+            const defaultRule = defaultRules.find((r) => r.rule_id === defaultRuleId);
+            if (defaultRule) {
+                logger.warn(`Adding default global ${defaultRule.rule_id}`);
```

Comment:
I suppose we should warn even louder if `defaultRule` is missing, as it suggests that a rule was added to `EXPECTED_DEFAULT_OVERRIDE_RULE_IDS` without an entry in `DEFAULT_OVERRIDE_RULES`?